physical/consul.go

```diff
@@ -577,3 +360,270 @@ func (c *ConsulLock) Value() (bool, string, error) {
+
+			// Abort if service discovery is disabled or a
+			// reconcile handler is already active
+			if !c.disableRegistration && atomic.CompareAndSwapInt64(&serviceRegLock, 0, 1) {
+				// Enter handler with serviceRegLock held
+				go func() {
+					defer atomic.CompareAndSwapInt64(&serviceRegLock, 1, 0)
+					for !shutdown {
+						serviceID, err := c.reconcileConsul(registeredServiceID, activeFunc, sealedFunc)
+						if err != nil {
+							c.logger.Printf("[WARN]: consul: reconcile unable to talk with Consul backend: %v", err)
+							time.Sleep(consulRetryInterval)
+							continue
+						}
+
+						c.serviceLock.Lock()
+						defer c.serviceLock.Unlock()
+
+						registeredServiceID = serviceID
+						return
+					}
+				}()
+			}
+		case <-checkTimer.C:
+			checkTimer.Reset(c.checkDuration())
+			// Abort if service discovery is disabled or a
+			// reconcile handler is active
+			if !c.disableRegistration && atomic.CompareAndSwapInt64(&checkLock, 0, 1) {
+				// Enter handler with serviceRegLock held
+				go func() {
+					defer atomic.CompareAndSwapInt64(&checkLock, 1, 0)
+					for !shutdown {
+						unsealed := sealedFunc()
+						if err := c.runCheck(unsealed); err != nil {
+							c.logger.Printf("[WARN]: consul: check unable to talk with Consul backend: %v", err)
+							time.Sleep(consulRetryInterval)
+							continue
+						}
+						return
+					}
+				}()
+			}
+		case <-shutdownCh:
+			c.logger.Printf("[INFO]: consul: Shutting down consul backend")
+			shutdown = true
+			break shutdown
+		}
+	}
+
+	c.serviceLock.RLock()
+	defer c.serviceLock.RUnlock()
+	if err := c.client.Agent().ServiceDeregister(registeredServiceID); err != nil {
+		c.logger.Printf("[WARN]: consul: service deregistration failed: %v", err)
+	}
+}
+
+// checkID returns the ID used for a Consul Check.  Assume at least a read
+// lock is held.
+func (c *ConsulBackend) checkID() string {
+	return "vault-sealed-check"
+}
+
+// reconcileConsul queries the state of Vault Core and Consul and fixes up
+// Consul's state according to what's in Vault.  reconcileConsul is called
+// without any locks held and can be run concurrently, therefore no changes
+// to ConsulBackend can be made in this method (i.e. wtb const receiver for
+// compiler enforced safety).
+func (c *ConsulBackend) reconcileConsul(registeredServiceID string, activeFunc activeFunction, sealedFunc sealedFunction) (serviceID string, err error) {
+	// Query vault Core for its current state
+	active := activeFunc()
+	sealed := sealedFunc()
+
+	agent := c.client.Agent()
+
+	// Get the current state of Vault from Consul
+	var currentVaultService *api.AgentService
+	if services, err := agent.Services(); err == nil {
+		if service, ok := services[c.serviceName]; ok {
+			currentVaultService = service
+		}
+	}
+
+	serviceID = c.serviceID()
+	tags := serviceTags(active)
+
+	var reregister bool
+	if currentVaultService == nil || registeredServiceID == "" {
```

Comment:
It certainly could be but I wanted to differentiate qualifying if we needed to reregister because we had an empty serviceID with comparing the values in an already registered service.  As a style note I've grown to break apart else/if statements out for future readers and leave the code folding to the compiler.